pkg/interceptors/interceptors.go

```diff
@@ -17,73 +17,84 @@ limitations under the License.
 package interceptors
 
 import (
-	"context"
+	"fmt"
 	"net/http"
-	"path"
+	"reflect"
+	"time"
 
 	triggersv1 "github.com/tektoncd/triggers/pkg/apis/triggers/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	fields "k8s.io/apimachinery/pkg/fields"
 	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
 )
 
-// Interceptor is the interface that all interceptors implement.
-type Interceptor interface {
-	ExecuteTrigger(req *http.Request) (*http.Response, error)
+// WebhookSecretStore provides cached lookups of k8s secrets, backed by a Reflector.
+type WebhookSecretStore interface {
+	Get(sr triggersv1.SecretRef) ([]byte, error)
 }
 
-type key string
+type webhookSecretStore struct {
+	store                  cache.Store
+	eventListenerNamespace string
+}
 
-const requestCacheKey key = "interceptors.RequestCache"
+// NewWebhookSecretStore provides cached lookups of k8s secrets, backed by a Reflector.
+func NewWebhookSecretStore(cs kubernetes.Interface, ns string, resyncInterval time.Duration, stopCh <-chan struct{}) WebhookSecretStore {
+	store := cache.NewStore(func(obj interface{}) (string, error) {
+		secret, ok := obj.(*corev1.Secret)
+		if !ok {
+			return "", fmt.Errorf("object is not a secret; got %s", reflect.TypeOf(obj))
+		}
 
-// WithCache clones the given request and sets the request context to include a cache.
-// This allows us to cache results from expensive operations and perform them just once
-// per each trigger.
-//
-// Each request should have its own cache, and those caches should expire once the request
-// is processed. For this reason, it's appropriate to store the cache on the request
-// context.
-func WithCache(req *http.Request) *http.Request {
-	return req.WithContext(context.WithValue(req.Context(), requestCacheKey, make(map[string]interface{})))
-}
+		return fmt.Sprintf("%s/%s", secret.Namespace, secret.Name), nil
+	})
 
-func getCache(req *http.Request) map[string]interface{} {
-	if cache, ok := req.Context().Value(requestCacheKey).(map[string]interface{}); ok {
-		return cache
+	secretStore := webhookSecretStore{
+		store:                  store,
+		eventListenerNamespace: ns,
 	}
 
-	return make(map[string]interface{})
+	lw := cache.NewListWatchFromClient(cs.CoreV1().RESTClient(), "secrets", metav1.NamespaceAll, fields.Everything())
```

Comment:
Using NamespaceAll here is probably  not a long term solution, since it requires cluster-wide list secrets in RBAC. I guess we could wrap a map of stores here - one per namespace referenced in calls to `Get` ...